sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileAsyncClient.java

```diff
@@ -148,6 +183,368 @@ public String getFileName() {
+                    fileOffset, length, headers, validatedUploadRequestConditions);
+
+            return createWithResponse(null, null, headers, metadata, requestConditions)
+                .then(UploadUtils.determineUploadFullOrChunked(data, validatedParallelTransferOptions,
+                uploadInChunksFunction, uploadFullMethod));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    private Mono<Response<PathInfo>> uploadInChunks(Flux<ByteBuffer> data, long fileOffset, long length,
+        ParallelTransferOptions parallelTransferOptions, PathHttpHeaders httpHeaders,
+        DataLakeRequestConditions requestConditions) {
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        AtomicLong totalProgress = new AtomicLong();
+        Lock progressLock = new ReentrantLock();
+
+        // Validation done in the constructor.
+        UploadBufferPool pool = new UploadBufferPool(parallelTransferOptions.getNumBuffers(),
+            parallelTransferOptions.getBlockSize(), MAX_APPEND_FILE_BYTES);
+
+        Flux<ByteBuffer> chunkedSource = UploadUtils.chunkSource(data, parallelTransferOptions);
+
+        /*
+         Write to the pool and upload the output.
+         */
+        return chunkedSource.concatMap(pool::write)
+            .concatWith(Flux.defer(pool::flush))
+            /* Map the data to a tuple, writing in the buffer.remaining temporarily */
+            .map(buffer -> Tuples.of(buffer, (long) buffer.remaining()))
+            /* The tuple keeps track of the next buffer to write and the fileOffset for the next buffer */
+            .scan((result, source) -> {
+                ByteBuffer buffer = source.getT1();
+                long currentBufferLength = buffer.remaining();
+                long lastBytesWritten = result.getT2();
+
+                return Tuples.of(buffer, currentBufferLength + lastBytesWritten);
+            })
+            .flatMapSequential(tuple2 -> {
+                ByteBuffer buffer = tuple2.getT1();
+                long currentBufferLength = buffer.remaining();
+                long currentOffset = tuple2.getT2() - currentBufferLength + fileOffset;
+                // Report progress as necessary.
+                Flux<ByteBuffer> progressData = ProgressReporter.addParallelProgressReporting(
+                    Flux.just(buffer), parallelTransferOptions.getProgressReceiver(), progressLock, totalProgress);
+                return appendWithResponse(progressData, currentOffset, currentBufferLength, null,
+                    requestConditions.getLeaseId())
+                    .doFinally(x -> pool.returnBuffer(buffer))
+                    .flux();
+            })
+            .last()
+            .flatMap(resp -> flushWithResponse(length, false, false, httpHeaders, requestConditions));
```

Comment:
I think `then()` is a bit of a cleaner way of doing this

Reply:
need to use flatmap now to pass length